src/planner/partialize.c

```diff
@@ -185,6 +197,657 @@ partialize_agg_paths(RelOptInfo *rel)
+							   double d_num_groups, GroupPathExtraData *extra_data)
+{
+	Query *parse = root->parse;
+
+	/* Determine costs for aggregations */
+	AggClauseCosts *agg_partial_costs = &extra_data->agg_partial_costs;
+
+	bool is_sorted = pathkeys_contained_in(root->group_pathkeys, path->pathkeys);
+
+	if (!is_sorted)
+	{
+		path = (Path *) create_sort_path(root, path->parent, path, root->group_pathkeys, -1.0);
+	}
+
+	Path *sorted_agg_path = (Path *) create_agg_path(root,
+													 path->parent,
+													 path,
+													 target,
+													 parse->groupClause ? AGG_SORTED : AGG_PLAIN,
+													 AGGSPLIT_INITIAL_SERIAL,
+													 parse->groupClause,
+													 NIL,
+													 agg_partial_costs,
+													 d_num_groups);
+
+	return sorted_agg_path;
+}
+
+/*
+ * Generate a partially hashed aggregated add path on top of a path
+ */
+static Path *
+create_hashed_partial_agg_path(PlannerInfo *root, Path *path, PathTarget *target,
+							   double d_num_groups, GroupPathExtraData *extra_data)
+{
+	Query *parse = root->parse;
+
+	/* Determine costs for aggregations */
+	AggClauseCosts *agg_partial_costs = &extra_data->agg_partial_costs;
+
+	Path *hash_path = (Path *) create_agg_path(root,
+											   path->parent,
+											   path,
+											   target,
+											   AGG_HASHED,
+											   AGGSPLIT_INITIAL_SERIAL,
+											   parse->groupClause,
+											   NIL,
+											   agg_partial_costs,
+											   d_num_groups);
+	return hash_path;
+}
+
+/*
+ * Add partially aggregated subpath
+ */
+static void
+add_partially_aggregated_subpaths(PlannerInfo *root, Path *parent_path,
+								  PathTarget *partial_grouping_target, double d_num_groups,
+								  GroupPathExtraData *extra_data, bool can_sort, bool can_hash,
+								  Path *subpath, List **sorted_paths, List **hashed_paths)
+{
+	/* Translate targetlist for partition */
+	AppendRelInfo *appinfo = ts_get_appendrelinfo(root, subpath->parent->relid, false);
+	PathTarget *chunktarget = copy_pathtarget(partial_grouping_target);
+	chunktarget->exprs =
+		castNode(List, adjust_appendrel_attrs(root, (Node *) chunktarget->exprs, 1, &appinfo));
+
+	/* In declarative partitioning planning, this is done by appy_scanjoin_target_to_path */
+	Assert(list_length(subpath->pathtarget->exprs) == list_length(parent_path->pathtarget->exprs));
+	subpath->pathtarget->sortgrouprefs = parent_path->pathtarget->sortgrouprefs;
+
+	if (can_sort)
+	{
+		*sorted_paths = lappend(*sorted_paths,
+								create_sorted_partial_agg_path(root,
+															   subpath,
+															   chunktarget,
+															   d_num_groups,
+															   extra_data));
+	}
+
+	if (can_hash)
+	{
+		*hashed_paths = lappend(*hashed_paths,
+								create_hashed_partial_agg_path(root,
+															   subpath,
+															   chunktarget,
+															   d_num_groups,
+															   extra_data));
+	}
+}
+
+/*
+ * Generate a total aggregation path for partial aggregations
```

Comment:
What is a "total aggregation" path? Is it one that includes the Gather node to make it "total" or is it a non-partial path?